### DISCUSSION TITLE
fail to expand field and update appstate (Dart / Flutterflow)

### DISCUSSION THREAD

**Discussion by WDuyv** [pocketbase/pocketbase#4872] [Q&A] [answered]
I'm pretty new to Dart, Flutterflow and Pocketbase, but so far I manage to write all the Dart code necessary for my FF actions.
But were I really fail is something that should be really straightforward, but somehow I'm stuck on it for days now, so I really hope someone here can help me.
 (P.s. if I  use 'recipeName: record.getStringValue('recipeName') I get the id).
 
 This is my code:

Future<void> listRecipes(String? userid) async {
  final pb = PocketBase('sample.com');

  final records = await pb.collection('Recipes').getFullList(
        sort: 'created',
        expand: 'Cookietype',
        filter: 'user.id="$userid"',
      );
  List<RecipeStruct> myList = [];
  records.forEach((record) {
    RecipeStruct recipeObject = RecipeStruct(
        id: record.id,
        recipeName: record.getStringValue('recipeName'),
        cookietype:
            record.expand['Cookietype']!.first.getStringValue('nameCookietype'));
    myList.add(recipeObject);
  });
  FFAppState().listRecipes = myList;

**Comment by ganigeorgiev** [comment]
It is not clear what is the question or the issue with the above code.

General info for expand you can find in https://pocketbase.io/docs/working-with-relations/#expanding-relations.

If you need further help please elaborate a little more what you are trying to do and export your collections configuration as json (Admin UI > Settings > Export collections).

**Reply by WDuyv** [reply]
Well, that's a bit the problem, as soon as I add:

_'expand: '_Cookietype'
cookietype: record.expand['Cookietype']!.first.getStringValue('nameCookietype'));_

to my code it doesn't return a value, but I also don't receive any error, so I'm a bit in the dark.
Thanks for the link, but of course that was the very first link I have been looking at, and I have been googling for days.
(Sorry I'm not going to export my collection)

**Reply by ganigeorgiev** [reply]
I'm still not sure what actually is the issue. If `record.expand['Cookietype']` is not set then double check that your client request satisfies the View API rule of the expanded collection.

Without details about your collections structure and fields I'm not sure how to help further.

**Reply by WDuyv** [reply]
So to you it also looks correct?
The log in PocketBase also shows no error (just info, get API).
I know the record and fieldnames are correct and have data.
These are the relevant pieces of the collection:

       "indexes": [],
        "listRule": "@request.auth.id = \"\"",
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    },
    {
        "id": "avlr0dd2q8bde00",
        "name": "Recipes",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "kczsajjl",
                "name": "recipeName",
                "type": "text",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "min": null,
                    "max": null,
                    "pattern": ""
                }
            },
            {
                "system": false,
                "id": "ofbl40xt",
                "name": "user",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": 1,
                    "displayFields": null
                }
            },
            {
                "system": false,
                "id": "0ajq49se",
                "name": "Cookietype",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "vmj3zcd9ioqx6je",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": 1,
                    "displayFields": null
                }
------------------------------------------------------------
   "indexes": [],
        "listRule": "@request.auth.id = recipe.user.id",
        "viewRule": "@request.auth.id = recipe.user.id",
        "createRule": "@request.auth.id = \"\"",
        "updateRule": "@request.auth.id = recipe.user.id",
        "deleteRule": "@request.auth.id = recipe.user.id",
        "options": {}
    },
    {
        "id": "vmj3zcd9ioqx6je",
        "name": "Cookietype",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "xnjute0g",
                "name": "nameCookietype",
                "type": "text",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "min": null,
                    "max": null,
                    "pattern": ""
                }

**Reply by ganigeorgiev** [reply]
Again, double check that your request satisfies the View API rule of the `Cookietype` collection, ~aka. that you are authenticated and that the authenticated user matches with the one from the recipe (_in your initial code snippet there is no `authWith*` call_).~

Actually the above collections structure export seems incomplete and there is no information about the API rules of the `Cookietype` collection to understand or attempt to reproduce it locally.

In any case, I don't think I'll be able to help you further.

If you are still not able to resolve your issue, then please clearly elaborate on your collection structure and what you are trying to do to avoid the unncessary back-and-forth, but until there is more information there is nothing here to discuss.

**Reply by WDuyv** [reply]
Ah too bad, I was hoping to get some suggestions here I could try out.
But this might be a Flutterflow issue.
I know it's not a authentication problem, because Pocketbase log would tell me, and even with both collections granting everyone access, the code without expand shows me the data, the code with expands doesn't come up with any data at all, not even the other fields.
I will take this issue to the FF community.
I think this has something to do with using 'expand' in a list, because I get the message 'The method 'getStringValue' isn't defined for the type 'List'' when I don't use 'first.getStringValue', which is kind of silly because in Pocketbase it is a single reference and I can use getStringValue for 'recipeName' also without any problem.

**Reply by WDuyv** [reply]
Ganigeorgiev, I found the problem, just no solution yet. I think there is something wrong with my authentication, and I have the feeling you might be able to help me with this.
I use authentication via e-mail/password, and that seems to work, I get a token.
If I set the API rules to @request.auth.id = "" I can access without problem, so at least that means I am authenticated, right?
If I change the API rules to @request.auth.id = user.id I can't access them, although I am logged in as that user and it was created by that user.
Then I took an other look at my log and I found:  All my requests are: data.auth : guest
Shouldn't I be seeing the user were it says 'guest'?

**Reply by ganigeorgiev** [reply, marked as answer]
I'm confused.

> If I set the API rules to @request.auth.id = "" I can access without problem, so at least that means I am authenticated, right?

No. When the auth id is equal to an empty string/null then this means that your rule allow only acces to non-authenticated users. If you want to grant access to only authenticated users you need to negate the operator - aka . `@request.auth.id != ""`.

> If I change the API rules to @request.auth.id = user.id I can't access them, although I am logged in as that user and it was created by that user.

Are you sure that you are actually logged in? From your initial code sample you are creating a new PocketBase instance before sending the request and `pb.authStore` in this case will not be initialized/loaded. For more details about the auth store you can check https://github.com/pocketbase/dart-sdk#authstore.

**Reply by WDuyv** [reply]
Thanks for your quick reply Ganigerogiev, I think you are right about not really being logged in.
I will look into this. I hope you don't mind if I post more questions here if I have them.

**Reply by WDuyv** [reply]
So I use an API call to log in, which works, but the handling after is done with settings within FlutterFlow (like persist authentication data , etc.). Even if I uncheck persist authentication, and I have to log in to the app again, as soon as I make a request to the database it is done as  guest.
You say:
'you are creating a new PocketBase instance before sending the request'. By that you mean the request for authentication right?
So I need to do something like?:

 // Authenticate first
  await pb.authenticate('your_api_key', 'your_token');
  
(Apologies for any dumb questions, I'm new to both Dart, PocketBase and API's)

**Reply by WDuyv** [reply]
Or like:

Future listRecipes(String? userid) async {
  final pb = PocketBase('sample.com', apiKey: 'YOUR_API_KEY', token: 'YOUR_TOKEN');

**Reply by ganigeorgiev** [reply]
> ... but the handling after is done with settings within FlutterFlow (like persist authentication data , etc.).

Unfortunately I'm not familiar with FlutterFlow and I'm not sure what is the recommended way to integrate with.

In general, you usually need only 1 global/singleton PocketBase SDK instance that you can reuse and inject in your app components. By default the Dart SDK's `pb.authStore` is not persistent, meaning that the auth state is kept only for the lifetime of the `pb` instance. For persistence (eg. in case the app restart) you need to load a custom store like SharedPreferences, Hive, local file, etc. as shown in the [README](https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#authstore):

```dart
final prefs = await SharedPreferences.getInstance();

final store = AsyncAuthStore(
 save:    (String data) async => prefs.setString('pb_auth', data),
 initial: prefs.getString('pb_auth'),
);

final pb = PocketBase('http://example.com', authStore: store);
```

If FlutterFlow has its own way to manage the app persistent and you want to store the token there, then you can also manually load the token (and optionally the auth model if serialized together) by doing something like:

```dart
pb.authStore.save('YOUR_TOKEN', optAuthModel)
```

**Reply by WDuyv** [reply]
Yeah I think I get the idea now.
I can store the token in the Appstate and then insert that into the header. Something like this:

// Get the authentication token from FFAppState
  String? token = FFAppState().userData.token;

  // Set the authorization header with the authentication token
  Map<String, String> headers = {
    'Authorization': 'Bearer $token',
  };
  
  I will post here again once I know for sure this works.
  thanks for your patience.

**Comment by WDuyv** [comment]
Yes, I finally get to see: data.auth | authRecord
Simply add the header. After months of coding I finally learning the basics....
Thanks ganigeorgiev!
